conan/cache/cache.py

```diff
@@ -51,60 +51,70 @@ def base_folder(self):
-        else:
-            return self._get_tmp_path()
-
-    def get_or_create_reference_layout(self, ref: ConanReference):
-        path = self._get_or_create_reference_path(ref)
-
-        if not ref.rrev:
-            ref = ConanReference(ref.name, ref.version, ref.user, ref.channel, RREV_UNKNOWN,
-                                 ref.pkgid, ref.prev)
-
-        reference_path, created = self._db.get_or_create_reference(path, ref)
-        self._create_path(reference_path, remove_contents=created)
-
+    @staticmethod
+    def _get_path(ref: ConanReference):
+        return md5(ref.full_reference)
+
+    def create_tmp_reference_layout(self, ref: ConanReference):
+        assert not ref.rrev, "Recipe revision should be unknown"
+        # even if they are temporal, we have to assign unique revisions to temporal references
+        # until we know the calculated revision so that we don't store multiple references
+        # with same rrev or prev
+        temp_rrev = f"{RREV_UNKNOWN}-{str(uuid.uuid4())}"
+        ref = ConanReference(ref.name, ref.version, ref.user, ref.channel, temp_rrev,
+                             ref.pkgid, ref.prev)
+        reference_path = self._get_tmp_path()
+        self._db.create_reference(reference_path, ref)
+        self._create_path(reference_path)
         return RecipeLayout(ref, os.path.join(self.base_folder, reference_path))
 
-    def get_or_create_package_layout(self, pref: ConanReference):
+    def create_tmp_package_layout(self, pref: ConanReference):
         assert pref.rrev, "Recipe revision must be known to get or create the package layout"
         assert pref.pkgid, "Package id must be known to get or create the package layout"
-        package_path = self._get_or_create_package_path(pref)
-        if not pref.prev:
-            pref = ConanReference(pref.name, pref.version, pref.user, pref.channel, pref.rrev,
-                                  pref.pkgid, PREV_UNKNOWN)
+        assert not pref.prev, "Package revision should be unknown"
+        temp_prev = f"{PREV_UNKNOWN}-{str(uuid.uuid4())}"
+        pref = ConanReference(pref.name, pref.version, pref.user, pref.channel, pref.rrev,
+                              pref.pkgid, temp_prev)
+        package_path = self._get_tmp_path()
+        self._db.create_reference(package_path, pref)
+        self._create_path(package_path)
+        return PackageLayout(pref, os.path.join(self.base_folder, package_path))
 
-        package_path, created = self._db.get_or_create_reference(package_path, pref)
-        self._create_path(package_path, remove_contents=created)
+    def create_reference_layout(self, ref: ConanReference):
+        assert ref.rrev, "Recipe revision must be known to create the package layout"
+        ref = ConanReference(ref.name, ref.version, ref.user, ref.channel, ref.rrev,
+                             ref.pkgid, ref.prev)
+        reference_path = self._get_path(ref)
+        self._db.get_or_create_reference(reference_path, ref)
```

Comment:
Maybe this should also be a ``db.create_reference``, to ensure nothing was there in the DB for this reference.